### PR TITLE
#164798916 User receive notifications

### PR DESCRIPTION
**Pull request by samorahno**
#### Description
Creates the function that forwards email notifications to the followers of an author immediately he publishes a new article.
#### Type of change
- [x]  Non-breaking change(introduces new functionality to codebase)
#### How Has This Been Tested?
All unit tests passed.

##### Screenshots
Screenshot 2019-04-08 at 10 23 18 AM
Screenshot 2019-04-08 at 10 23 37 AM
Screenshot 2019-04-08 at 10 23 54 AM

#### Checklist:
- [x]  Modify `users` model to include `subscribed`
- [x] Create `followers` model
- [x] Create `notifications` model
- [x] Seed `followers` and `users` model with data required for unit tests
- [x] Write tests
- [x] Write function to process email notifications for new article
- [x] Write function to process email notifications for new follower
- [x] Seed articles table
- [x] Write function to process email notifications for new comment
#### PIVOTAL BOARD
[#164798916](https://www.pivotaltracker.com/story/show/164798916)

### CODE REVIEW

File: server/migrations/20190404074536-create-follower.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+          key: 'id',
+          as: 'follower',
+        }
+      },
+    });
+  },
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.dropTable('Followers');
+  }
+};
```
Comment (waleCloud) [inline server/migrations/20190404074536-create-follower.js]: newline here

File: server/migrations/20190404082058-create-notification.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+          model: 'Users',
+          key: 'id',
+        }
+      }
+    });
+  },
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.dropTable('Notifications');
+  }
+};
```
Comment (waleCloud) [inline server/migrations/20190404082058-create-notification.js]: Kindly fix this newline

File: server/models/notification.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    },
+  }, {});
+  Notification.associate = (models) => {
+    // associations can be defined here
+    Notification.belongsTo(models.User, {
+      foreignKey: 'userId',
+    });
+  };
+  return Notification;
+};
```
Comment (waleCloud) [inline server/models/notification.js]: Is hound no longer in the repo?

**Review by waleCloud — CHANGES_REQUESTED** [review]

File: server/utils/notifications/followers-article-mail.util.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  const followers = await user.getFollowers({
+    attributes: ['id', 'email', 'firstname', 'lastname'],
+    where: { subscribed: true }
+  });
+
+  if (followers.length === 0) {
+    return;
+  }
+
+  followers.forEach(async (follower) => {
```
Comment (waleCloud) [inline server/utils/notifications/followers-article-mail.util.js]: In a situation where the followers are so much, how does this affect performance?
Is there a better way to solve for this aside from sending emails in a loop.

File: server/migrations/20190404074536-create-follower.js (hunk 1 above)
Comment (samorahno) [reply server/migrations/20190404074536-create-follower.js]: @waleCloud sorry about this, we will add the new lines right away.

File: server/utils/notifications/followers-article-mail.util.js (hunk 4 above)
Comment (chukwudidu) [reply server/utils/notifications/followers-article-mail.util.js]: @waleCloud we realized the implication of using a loop and tabled two options to @paradisekelechi. One was to loop through and the other, to use cron jobs to process the notifications in batches. The major drawback of using cron jobs would be that the notifications would not be instantaneous so we agreed to make do with looping for now.

But as the application scales, however, we would need to implement a more optimal solution.

File: server/models/notification.js (hunk 3 above)
Comment (chukwudidu) [reply server/models/notification.js]: It is. I am not sure as to whether it has stopped working. Would look into this.

Comment (chukwudidu) [reply server/models/notification.js]: It is actually working. Hound just picked up some violation.

Comment (BukkyOmo) [reply server/models/notification.js]: You still haven't fixed the end of line issue.

**Comment by paradisekelechi** [comment]
Handle the conflicts and hound issues

File: server/utils/notifications/follows-mail.util.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ * @param {int} followeeId user being followed
+ * @param {int} followerId user who followed
+ * @returns null
+ */
+
+const newFollowerMail = async (followeeId, followerId) => {
+  const followee = await User.findByPk(followeeId);
+  const follower = await User.findByPk(followerId);
+
+  if (!followee || !follower) {
```
Comment (geekdanitek) [inline server/utils/notifications/follows-mail.util.js:18]: Good job on checking who will receive notifications.

**Review by geekdanitek — COMMENTED** [review]
The mail output design look good.

File: .eslintrc [hunk 6]
```diff
@@ -45,4 +46,4 @@
         ],
         "require-jsdoc": 0
     }
-}
+}
```
Comment (paradisekelechi) [inline .eslintrc:49]: Newline error message

File: server/utils/notifications/article-mail.util.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Process new article email notification to followers
+ * @param {int} userId id of the author
+ * @param {string} articleTitle
+ * @param {int} articleId
+ * @returns null
+ */
+
+const newArticleMail = async (userId, articleTitle, articleId) => {
```
[the remark this answers is not in the document]
Comment (paradisekelechi) [reply server/utils/notifications/article-mail.util.js:15]: This hound issue still exists

File: server/utils/notifications/comment-mail.util.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+const { User, Notification, Article } = db;
+
+/**
+ * Process new comment email notification to author of an article
+ * @param {int} articleId
+ * @param {int} userId user who commented
+ * @returns null
+ */
+
+const newCommentMail = async (articleId, userId) => {
```
[the remark this answers is not in the document]
Comment (paradisekelechi) [reply server/utils/notifications/comment-mail.util.js:14]: Hound issue!

File: server/utils/notifications/follows-mail.util.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+const { User, Notification } = db;
+
+/**
+ * Process email notification to a user who has been followed
+ * @param {int} followeeId user being followed
+ * @param {int} followerId user who followed
+ * @returns null
+ */
+
+const newFollowerMail = async (followeeId, followerId) => {
```
[the remark this answers is not in the document]
Comment (paradisekelechi) [reply server/utils/notifications/follows-mail.util.js:14]: Hound again

**Review by paradisekelechi — CHANGES_REQUESTED** [review]

File: .eslintrc (hunk 6 above)
Comment (chukwudidu) [reply .eslintrc:49]: > Handle the conflicts and hound issues

@paradisekelechi all done.

**Pull request merged**